cdap-tms/src/main/java/io/cdap/cdap/messaging/subscriber/AbstractMessagingSubscriberService.java

```diff
@@ -131,74 +160,114 @@ protected final String loadMessageId() {
+   *  Process messages in multiple transactions where each transaction has size of {@link this#txSize} messages
+   *  or size of ONE in case {@link this#shouldRunInSeparateTx(ImmutablePair)} returns true
+   *  For example, for given (m1, m2, m3, m4, m5) with txSize equals 3, and  shouldRunInSeparateTx(m3)=true
+   *  we process them in the following three transactions (m1, m2) (m3) (m4, m5)
+   * @param messages an {@link Iterable} of {@link ImmutablePair}, with the {@link ImmutablePair#first}
+   *                 as the message id, and the {@link ImmutablePair#second} as the decoded message
+   * @return
+   * @throws Exception
+   */
   @Nullable
   @Override
   protected String processMessages(Iterable<ImmutablePair<String, T>> messages) throws Exception {
-    MessageTrackingIterator iterator;
+    List<ImmutablePair<String, T>> currentTxMessages;
+    String lastMessageId = null;
+    MessageTrackingIterator iterator = new MessageTrackingIterator(messages.iterator());
+    Stopwatch stopwatch = new Stopwatch();
+    while (iterator.hasNext()) {
+      currentTxMessages = new ArrayList<>();
 
-    // Process the notifications and record the message id of where the processing is up to.
-    // 90% of the tx timeout is .9 * 1000 * txTimeoutSeconds = 900 * txTimeoutSeconds
-    long timeBoundMillis = 900L * txTimeoutSeconds;
-    iterator = TransactionRunners.run(getTransactionRunner(), context -> {
-      TimeBoundIterator<ImmutablePair<String, T>> timeBoundMessages = new TimeBoundIterator<>(messages.iterator(),
-                                                                                              timeBoundMillis);
-      MessageTrackingIterator trackingIterator = new MessageTrackingIterator(timeBoundMessages);
-      processMessages(context, trackingIterator);
-      String lastMessageId = trackingIterator.getLastMessageId();
+      while (iterator.hasNext() && currentTxMessages.size() < this.txSize) {
+        if (shouldRunInSeparateTx(iterator.peek())) {
+          // If next message should run in a separate transaction, two cases exist:
+          // 1. if currentTxMessages is empty, we can add next message and go to process it.
+          // 2. If currentTxMessages is not empty, we have to process what is in currentTxMessages, and
+          // process next message in the next round.
+          if (currentTxMessages.isEmpty()) {
+            currentTxMessages.add(iterator.next());
+          }
+          break;
+        }
+        currentTxMessages.add(iterator.next());
+      }
+
+      try {
+        String result = processSingleTxn(currentTxMessages, stopwatch);
+        if (result != null) {
+          // we only set lastMessageId if result of current batch is not null
+          // in order to avoid setting lastMessageId to null in case current batch returned null
+          // but previous batch finished.
+          lastMessageId = result;
+        }
+      } catch (Exception ex) {
+        if (lastMessageId == null) {
```

Comment:
may be good to log if the last message id is not null

Reply:
Done